src/compiler/checker.ts

```diff
@@ -13663,8 +13663,33 @@ namespace ts {
         // Return the contextual type for a given expression node. During overload resolution, a contextual type may temporarily
         // be "pushed" onto a node using the contextualType property.
         function getApparentTypeOfContextualType(node: Expression): Type {
-            const type = getContextualType(node);
-            return type && getApparentType(type);
+            let contextualType = getContextualType(node);
+            contextualType = contextualType && getApparentType(contextualType);
+            if (contextualType && contextualType.flags & TypeFlags.Union && isObjectLiteralExpression(node)) {
+                let match: Type | undefined;
+                propLoop: for (const prop of node.properties) {

```

Comment:
this code is *almost* exactly the same as the also-highly-questionable `findMatchingDiscriminantType`. The only real difference, I suspect, is that this code doesn't bail if `(type === match)`. Two things:

1. Can you harmonise the two code paths so there's only one `findMatchingDiscriminantType`? 
2.It's quite possible that the excess-property usage and union-error usage would both benefit from the `(type === match)` addition, so it could be that you can just swap out the existing body for this code..
3. Three things! Using `findMatchingDiscriminantType` is slow, and only makes the compiler faster because it eliminates unions early on. Can you test performance with this change?

Reply:
1. `findMatchingDiscriminantType` operates on type members, this operates on an actual node tree (as most contextual type operations do), so sadly I don't think they can be merged.
2. Sure, I'll try it; but I don't think it'll show up in much unless we add a test case where multiple fields are capable of acting as a discriminant for an object and discriminate to the same types. 🐱
3. Sure, will do.

Reply:
Oh hey, waddoyaknow, we already do have a test (`excessPropertyCheckWithUnions`) improved by changing `findMatchingDiscriminantType`. Neat.